### PR TITLE
Add proper types to runHandler

### PR DESCRIPTION
**Pull request by twk3**
## Summary

- Args and return values from runHandler should now work within loot core
- Updated some of the handler types to be more accurate

Split from https://github.com/actualbudget/actual/pull/2053 to start preparing to bring types to the API.

## Verify

Can be verified by going to one of the places that runHandlers is called, like the main.test.ts or the mocks/budget.ts and trying to put in some incorrect args to a handler that is strongly typed.

Like the following where a `test: "this will error"` was added:

```
await runHandler(handlers['schedule/create'], {
        schedule: {
          name: 'Phone bills',
          posts_transaction: false,
          test: "this will error",
        },
        conditions: [
          {
            op: 'is',
            field: 'payee',
            value: payees.find(item => item.name === 'Dominion Power').id,
          },
          {
            op: 'is',
            field: 'account',
            value: account.id,
          },
          {
            op: 'is',
            field: 'date',
            value: {
              start: monthUtils.currentDay(),
              frequency: 'monthly',
              patterns: [],
              skipWeekend: false,
              weekendSolveMode: 'after',
            },
          },
          { op: 'isapprox', field: 'amount', value: -12000 },
        ],
      });
```

Then either `yarn typecheck` or your IDE with flag the error.

### CODE REVIEW

File: packages/loot-core/src/types/server-handlers.d.ts [hunk 1]
```diff
@@ -332,10 +333,7 @@ export interface ServerHandlers {
 
   'close-budget': () => Promise<'ok'>;
 
-  'delete-budget': (arg: {
-    id?: string;
-    cloudFileId?: string;
-  }) => Promise<'ok'>;
+  'delete-budget': (arg: { id; cloudFileId? }) => Promise<'ok'>;
```
Comment (twk3) [inline packages/loot-core/src/types/server-handlers.d.ts:336]: whoops, this looks like a regression. I'll get this restored.

File: packages/loot-core/src/types/server-handlers.d.ts [hunk 2]
```diff
@@ -283,7 +283,8 @@ export interface ServerHandlers {
   'subscribe-needs-bootstrap': (
     args: { url } = {},
   ) => Promise<
-    { error: string } | { bootstrapped: unknown; hasServer: boolean }
+    | { error: string }
+    | { bootstrapped: unknown; hasServer: boolean; error?: undefined }
```
Comment (MatissJanis) [inline packages/loot-core/src/types/server-handlers.d.ts:287]: **💬 suggestion:** ‏I don't think we need `error` here
```suggestion
    | { bootstrapped: unknown; hasServer: boolean; }
```

**Review by MatissJanis — COMMENTED** [review]
Nice! Thanks for hardening our types

File: packages/loot-core/src/types/server-handlers.d.ts (hunk 2 above)
Comment (twk3) [reply packages/loot-core/src/types/server-handlers.d.ts:287]: @MatissJanis 
We currently have a deconstruction of the return of this handler here: https://github.com/actualbudget/actual/blob/d8996405c43a8b1dd530ccceeb78fe049e03919e/packages/loot-core/src/server/main.ts#L1673

typescript won't let you use deconstruction on an object with union types unless the properties you are deconstructing exist on all types in the union https://github.com/microsoft/TypeScript/issues/12815

wasn't sure whether we wanted to drop the deconstruction in favour of a result object, or adjust the typescript to allow us to keep the deconstruction. Thoughts?

Comment (MatissJanis) [reply packages/loot-core/src/types/server-handlers.d.ts:287]: The ultimate source of truth should be [the handler](https://github.com/actualbudget/actual/blob/d8996405c43a8b1dd530ccceeb78fe049e03919e/packages/loot-core/src/server/main.ts#L1519-L1521) itself. Which never returns `error` together with the other properties.

So in that sense: the consumer (destruction) is wrong and should be changed. Would you be open to changing it slightly?

File: packages/loot-core/src/server/main.ts [hunk 3]
```diff
@@ -1670,14 +1670,11 @@ handlers['set-server-url'] = async function ({ url, validate = true }) {
...
-        {
-          url,
-        },
-      );
-      if (error) {
-        return { error };
+      const result = await runHandler(handlers['subscribe-needs-bootstrap'], {
+        url,
+      });
+      if ('error' in result) {
```
Comment (twk3) [inline packages/loot-core/src/server/main.ts:1676]: This is what we have to do to type assert the union being returned from `handlers['subscribe-needs-bootstrap']`

**Comment by twk3** [comment]
@MatissJanis this is ready for another review

**Review by MatissJanis — DISMISSED** [review]
Perfect, thanks!

**Review by MatissJanis — DISMISSED** [review]

**Pull request merged**